spmd/api.py

```diff
@@ -0,0 +1,318 @@
+    kwargs = node.kwargs
+
+    # run dispatch once to get the real DTensor output
+    out = operator_dispatch(
+        node.target,
+        args,
+        node.kwargs,  # kwargs in this set of tests are all constants
+        DTensor._op_to_rules,
+        DTensor._custom_dispatch_ops,
+    )
+    node_to_obj[node] = out
+
+    # get DTensor specs for inputs and outputs
+    (target_schema, redistribute, output_sharding,) = propagate_input_sharding(
+        node.target,
+        args,
+        kwargs,
+        DTensor._op_to_rules,
+    )
+
+    flatten_args, args_tree_spec = tree_flatten(args)
+    flatten_args_schema, _ = tree_flatten(target_schema.args_schema)
+
+    specs: Dict[
+        torch.Tensor,
+        Tuple[
+            torch.Size,
+            DeviceMesh,
+            Sequence[Placement],
+            Sequence[Placement],
+        ],
+    ] = {}
+    for i, arg in enumerate(flatten_args):
+        if isinstance(arg, DTensor) and redistribute:
+            specs[arg._local_tensor] = (
+                arg.size(),
+                flatten_args_schema[i].mesh,
+                arg.placements,
+                flatten_args_schema[i].placements,
+            )
+
+    dispatch = partial(
+        _dispatch_with_local_tensors,
+        node.target,
+        kwargs=kwargs,
+        specs=specs,
+    )
+
+    def unwrap_local(e):
+        return e._local_tensor if isinstance(e, DTensor) else e
+
+    return make_fx(dispatch)(tree_map(unwrap_local, args))
+
+
+def _convert_to_distributed(
+    gm: fx.Graph,
+    inps: List[torch.Tensor],
+    schemas: List[Schema],
+    _allow_partial: bool = False,
+) -> fx.Graph:
+    node_to_obj: Dict[fx.Node, object] = {}
+    # map local op node in traced_f to its corresponding subgraph of
+    # DTensor ops.
+    replacements: Dict[torch.fx.Node, torch.fx.GraphModule] = {}
+
+    for i, node in enumerate(gm.graph.nodes):
+        if node.op == "placeholder":
+            assert i < len(
+                inps
+            ), f"got more placeholer nodes ({i + 1}) than inputs ({len(inps)})"
+            node_to_obj[node] = DTensor.from_local(
+                inps[i], schemas[i].mesh, schemas[i].placements
+            )
+        elif isinstance(node.target, torch._ops.OpOverload):
+            replacements[node] = _get_dtensor_dispatch_graph(node, node_to_obj)
+        elif node.op == "output":
+            if not _allow_partial:
+                new_args = []
+                has_partial = False
+                for a in node.args[0]:
+                    if not isinstance(a, fx.Node):
+                        new_args.append(a)
+                        continue
+                    obj = node_to_obj[a]
+                    if isinstance(obj, DTensor) and isinstance(
+                        obj.placements[0], _Partial
+                    ):
+                        has_partial = True
+
+                        def dummy_add(grad, zero) -> torch.Tensor:
```

Comment:
Q: this code is mainly to trigger allreduce for partial -> replicate conversion, so that we don't return partial tensors after the backward pass on `nn.Module`. Let me know if we no longer want a module wrapper in tau repo

Reply:
iirc we couldn't trace a train step because aot does not consider optimizer step into its min cut algorithm? My view on this is that it might make sense to just group backward graph + optimizer graph together as lots of optimizer is done with these two together. cc @Chillee 

Let's explicitly trigger all_reduce for now to unblock, and we can discuss more what might work best.